Source/Core/AbsyCmd.cs

```diff
@@ -3762,6 +3766,8 @@ public object ErrorData
       set { errorData = value; }
     }
 
+    public ProofObligationDescription Description { get; set; } = new AssertionDescription();
```

Comment:
Shouldn't this be `static`, there's no interface being implemented right?

Reply:
There's no interface being implemented, but the idea is that some clients may pass in something other than `AssertionDescription`, so not all `AssertCmd` instances will have the same `Description`.

Reply:
How would clients pass in something different? I see that this field has a public setter but at what point would you assign using that? Could you add a unit test that shows how to use the new API ?

To me it seems counterintuitive to have metadata of a type be held in an instance field of said type, as opposed to it being a virtual property/method or a static one.

Reply:
Although the Boogie codebase currently has only one type for descriptions of an `AssertCmd`, other language clients will likely have many. So different instances of `AssertCmd` (for instance), will have different `Description` properties, instead of the current case where they have different `ErrorData` values.

Said another way, it's not metadata about a type (even though there's currently only one description appropriate for a given type in the Boogie codebase). It's metadata about an instance of that type, which could be instantiated any way a client chooses. Clients will customize `AssertCmd`s by subclassing `AssertionDescription` rather than by subclassing `AssertCmd`.

Reply:
You can see [this commit](https://github.com/atomb/dafny/commit/ad105bbbe3516cfc9aba90472e993dcfc475f069), that I haven't yet turned into a PR, for an example of how that'd work.

Reply:
To me it looks like the current API is:

```csharp
// library
record AssertCmd(ProofObligationDescription description) : PredicateCmd {

}

abstract class ProofObligationDescription {
  public abstract string SuccessDescription { get; }
  ...
}

class DefaultAssertDescription {
  public override string SuccessDescription => "foo";
}

// consumer
class SpecificAssertDescription {
  public override string SuccessDescription => "defaultSuccessDescription";
}
class User {
  void Use() {
    var defaultAssertCmd = new AssertCmd(new DefaultAssertDescription());
    var specificAssertCmd = new AssertCmd(new SpecificAssertDescription());
  }
}
```

and my suggestion is to change that to:

```csharp
// library
abstract class AssertCmd : PredicateCmd, ProofObligation {
  public string SuccessDescription => "defaultSuccessDescription";
}

// this interface is optional. 
interface ProofObligation {
  string SuccessDescription { get; }
  ...
}

// consumer
class SpecificAssertCmd : AssertCmd {
  public override string SuccessDescription => "foo";
}

class User {
  void Use() {
    var defaultAssertCmd = new AssertCmd();
    var specificAssertCmd = new SpecificAssertCmd();
  }
}
```

The reason I prefer the latter version is that in this PR, there's only 1 specific value of `ProofObligationDescription` passed to AssertCmd, so it seems strange to have an instance field for that. Yes you want to be able to customise this in consumers, but don't need a field for that since you can also use subclassing.

Reply:
That is another reasonable design approach. There are a few reasons I prefer an additional constructor parameter over subclassing in this case.

The reason I most want to emphasize is that, to me, subclassing Boogie commands feels like extending the language rather than just adding metadata (even if the effect in this particular case is the same). I feel like Boogie is more coherent as a compilation target language than as a library that can be extended in arbitrary ways. Attaching complex, structured metadata to compilation targets, on the other hand, is a common thing (see LLVM), and frequently quite useful (as in this case).

Another reason is that the approach you're suggesting widens the interface of commands (at least client-specific subclasses). It would add, say, 3 properties to each derived command, rather than one property with a value that has several properties of its own. Grouping properties like that tends to lead to more general-purpose code and easier-to-learn interfaces, in my experience.

Reply:
> I feel like Boogie is more coherent as a compilation target language than as a library that can be extended in arbitrary ways.

My impression is that the API this PR introduces cannot be used through a .bpl program, so it requires consuming Boogie as a library anyways.

> The reason I most want to emphasize is that, to me, subclassing Boogie commands feels like extending the language rather than just adding metadata (even if the effect in this particular case is the same). ... 
Attaching complex, structured metadata to compilation targets, on the other hand, is a common thing (see LLVM), and frequently quite useful (as in this case).

I can see how you feel that. Let's try avoiding the subclasses then. Also, do you happen to have a link to an LLVM example of this?

One more thing, it seems that Boogie uses subclasses of AssertCmd to configure error messages. Would it make sense to stop doing that and use the instance field this PR adds instead? After this PR there's some duplication of classes, for example there's `InvariantEstablishedDescription` and `LoopInitAssertCmd`. Do we still need `LoopInitAssertCmd` now that we have `InvariantEstablishedDescription` ?